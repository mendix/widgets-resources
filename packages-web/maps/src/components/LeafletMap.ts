import { Component, createElement } from "react";
import {
    FeatureGroup,
    LatLngLiteral,
    LeafletEvent,
    Map,
    Marker,
    TileLayerOptions,
    icon,
    tileLayer
} from "leaflet";
import * as classNames from "classnames";
import ReactResizeDetector from "react-resize-detector";

import { Container, MapUtils } from "../utils/namespace";
import Utils from "../utils/Utils";
import { Alert } from "./Alert";
type MapProps = Container.MapProps;
type DataSourceLocationProps = Container.DataSourceLocationProps;
type Location = Container.Location;
type SharedProps = MapUtils.SharedProps;

const customUrls = Utils.customUrls;
const mapAttr = Utils.mapAttr;

export interface LeafletMapProps extends SharedProps, MapProps {
    onClickMarker?: (event: LeafletEvent, locationAttr: DataSourceLocationProps) => void;
}

export interface LeafletMapState {
    center: LatLngLiteral;
    alertMessage?: string;
    resized: boolean;
}

export class LeafletMap extends Component<LeafletMapProps, LeafletMapState> {
    private leafletNode?: HTMLDivElement;
    private defaultCenterLocation: LatLngLiteral = { lat: 51.9107963, lng: 4.4789878 };
    private map?: Map;
    private markerGroup = new FeatureGroup();
    private readonly onResizeHandle = this.onResize.bind(this);

    readonly state: LeafletMapState = {
        center: this.defaultCenterLocation,
        resized: false
    };

    render() {
        return createElement("div", {
                className: classNames("widget-maps", this.props.className),
                style: { ...this.props.divStyles , ...Utils.getDimensions(this.props) }
            },
            createElement(Alert, {
                bootstrapStyle: "danger",
                className: "widget-leaflet-maps-alert leaflet-control"
            }, this.state.alertMessage),
            createElement("div",
                {
                    className: "widget-leaflet-maps-wrapper"
                },
                createElement("div", {
                    className: "widget-leaflet-maps",
                    ref: this.getRef
                }),
                createElement(ReactResizeDetector, {
                    handleWidth: true,
                    handleHeight: true,
                    refreshRate: 100,
                    refreshMode: "throttle",
                    onResize: this.onResizeHandle
                })
            )
        );
    }

    componentDidMount() {
        if (this.leafletNode && !this.map) {
            this.map = new Map(this.leafletNode, {
                    scrollWheelZoom: this.props.optionScroll,
                    zoomControl: this.props.optionZoomControl,
                    attributionControl: this.props.attributionControl,
                    zoom: this.props.zoomLevel,
                    minZoom: 2,
                    // Work around for page scroll down to botom on first click on map in Chrome and IE
                    // https://github.com/Leaflet/Leaflet/issues/5392
                    keyboard: false,
                    dragging: this.props.optionDrag
                })
                .addLayer(this.setTileLayer());
            if (this.props.inPreviewMode) {
                this.setDefaultCenter(this.props);
            }
        }
    }

    componentWillReceiveProps(nextProps: LeafletMapProps) {
        if (nextProps.alertMessage !== this.props.alertMessage) {
            this.setState({ alertMessage: nextProps.alertMessage });
        }
        if (this.props.allLocations !== nextProps.allLocations) {
            this.setDefaultCenter(nextProps);
            this.setState({ resized: false });
        }
    }

    componentDidUpdate() {
        if (this.map && !this.props.fetchingData) {
            this.map.panTo(this.state.center);
        }
    }

    componentWillUnmount() {
        if (this.map) {
            this.map.remove();
        }
    }

    private onResize() {
        // When map is placed on the non default tab, the maps has no size.
        // Therefor it need to update on the first view.
        if (this.map && !this.state.resized) {
            this.setDefaultCenter(this.props);
            this.setBounds();
            this.setState({ resized: true });
        }
    }

    private getRef = (node: HTMLDivElement) => {
        this.leafletNode = node;
    }

    private setTileLayer = () => {
        const { mapProvider, mapsToken } = this.props;

        let urlTemplate = "";
        let mapAttribution = "";
        if (mapProvider === "mapBox") {
            urlTemplate = customUrls.mapbox + mapsToken;
            mapAttribution = mapAttr.mapboxAttr;
        } else if (mapProvider === "hereMaps" && mapsToken && mapsToken.indexOf(",") > 0) {
            const splitToken = mapsToken.split(",");
            urlTemplate = customUrls.hereMaps + `app_id=${splitToken[0]}&app_code=${splitToken[1]}`;
            mapAttribution = mapAttr.hereMapsAttr;
        } else {
            urlTemplate = customUrls.openStreetMap;
            mapAttribution = mapAttr.openStreetMapAttr;
        }

        return tileLayer(urlTemplate, {
            attribution: mapAttribution,
            id: mapProvider === "mapBox" ? "mapbox.streets" : undefined
        } as TileLayerOptions);
    }

    private setDefaultCenter = (props: LeafletMapProps) => {
        if (!props.fetchingData) {
            this.renderMarkers(props.allLocations);
        }
    }

    private renderMarkers = (locations?: Location[]) => {
        this.markerGroup.clearLayers();
        if (locations && locations.length) {
            locations.forEach(location => this.createMarker(location).then(marker =>
                this.markerGroup.addLayer(marker.on("click", event => {
                    if (this.props.onClickMarker && location.locationAttr && location.locationAttr.onClickEvent !== "doNothing") {
                        this.props.onClickMarker(event, location.locationAttr);
                    }
                }))).then(layer => this.map ? this.map.addLayer(layer) : undefined)
                .catch(reason => this.setState({ alertMessage: `${reason}` }))
            );
            this.setBounds();
        } else if (this.map) {
            this.map.removeLayer(this.markerGroup);
        }
    }

    private setBounds = () => {
        const { defaultCenterLatitude, defaultCenterLongitude, autoZoom, zoomLevel } = this.props;
        setTimeout(() => {
            if (this.map && this.markerGroup) {
                const bounds = this.markerGroup.getBounds();
                if (bounds.isValid()) {
                    this.map.fitBounds(this.markerGroup.getBounds(), { animate: false }).invalidateSize();
                }
                if (!autoZoom) {
                    if (defaultCenterLatitude && defaultCenterLongitude) {
                        this.map.panTo({
                            lat: parseFloat(defaultCenterLatitude),
                            lng: parseFloat(defaultCenterLongitude)
                        }, { animate: false });
                    }
                    this.map.setZoom(zoomLevel);
                }
            }
        }, 0);
    }

    private createMarker = (location: Location): Promise<Marker> =>
        new Promise((resolve) => {
            const { latitude, longitude, url } = location;
            if (url) {
                resolve(
                    new Marker([ Number(latitude), Number(longitude) ]).setIcon(icon({
                        iconUrl: url,
                        iconSize: [ 38, 95 ],
                        iconAnchor: [ 22, 94 ],
                        className: "marker"
                    }))
                );
            } else {
                resolve(new Marker([ Number(latitude), Number(longitude) ]));
            }
        })
}
