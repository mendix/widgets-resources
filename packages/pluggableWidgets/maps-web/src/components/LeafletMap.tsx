import { createElement, ReactElement, useEffect, useRef } from "react";
import { Map, Marker as MarkerComponent, Popup, TileLayer } from "react-leaflet";
import classNames from "classnames";
import { SharedProps } from "../../typings/shared";
import { MapProviderEnum } from "../../typings/MapsProps";
import { getDimensions } from "../utils/dimension";
import { translateZoom } from "../utils/zoom";
import { latLngBounds } from "leaflet";
import { baseMapLayer } from "../utils/leaflet";

export interface LeafletProps extends SharedProps {
    mapProvider: MapProviderEnum;
    attributionControl: boolean;
}

export function LeafletMap(props: LeafletProps): ReactElement {
    const map = useRef<Map>();
    const center = { lat: 51.906688, lng: 4.48837 };
    const {
        autoZoom,
        attributionControl,
        className,
        currentLocation,
        locations,
        mapProvider,
        mapsToken,
        optionScroll: scrollWheelZoom,
        optionZoomControl: zoomControl,
        style,
        zoomLevel: zoom,
        optionDrag: dragging
    } = props;

    useEffect(() => {
        if (map.current) {
            const { leafletElement: mapRef } = map.current;
            const bounds = latLngBounds(
                locations
                    .concat(currentLocation ? [currentLocation] : [])
                    .filter(m => !!m)
                    .map(m => [m.latitude, m.longitude])
            );
            if (bounds.isValid()) {
                if (autoZoom) {
                    mapRef.fitBounds(bounds, { padding: [0.5, 0.5] }).invalidateSize();
                } else {
                    mapRef.panTo(bounds.getCenter(), { animate: false });
                }
            }
        }
    }, [map.current, locations, currentLocation, autoZoom]);

    return (
        <div className={classNames("widget-maps", className)} style={{ ...style, ...getDimensions(props) }}>
            <div className="widget-leaflet-maps-wrapper">
                <Map
                    attributionControl={attributionControl}
                    center={center}
                    className="widget-leaflet-maps"
                    dragging={dragging}
                    maxZoom={18}
                    minZoom={1}
                    ref={ref => {
                        if (ref && ref !== map.current) {
                            map.current = ref;
                        }
                    }}
                    scrollWheelZoom={scrollWheelZoom}
                    zoom={autoZoom ? translateZoom("city") : zoom}
                    zoomControl={zoomControl}
                >
                    <TileLayer
                        {...baseMapLayer(mapProvider, mapsToken)}
                        id={mapProvider === "mapBox" ? "mapbox.streets" : undefined}
                    />
                    {locations
                        .concat(currentLocation ? [currentLocation] : [])
                        .filter(m => !!m)
                        .map((marker, index) => (
                            <MarkerComponent
                                key={`marker_${index}`}
                                position={{ lat: marker.latitude, lng: marker.longitude }}
                                onclick={marker.title ? undefined : marker.onClick}
                                interactive={!!marker.title || !!marker.onClick}
                                title={marker.title}
                            >
                                {marker.title && (
                                    <Popup>
                                        <span
                                            style={{ cursor: marker.onClick ? "pointer" : "none" }}
                                            onClick={marker.onClick}
                                        >
                                            {marker.title}
                                        </span>
                                    </Popup>
                                )}
                            </MarkerComponent>
                        ))}
                </Map>
            </div>
        </div>
    );
}
