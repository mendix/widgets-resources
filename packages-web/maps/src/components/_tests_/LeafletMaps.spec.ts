import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";

import { LeafletMap, LeafletMapProps } from "../LeafletMap";

describe("Leaflet maps", () => {
    const defaultProps: LeafletMapProps = {
        autoZoom: true,
        defaultCenterLatitude: "",
        defaultCenterLongitude: "",
        height: 75,
        heightUnit: "pixels",
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        zoomLevel: 10,
        width: 50,
        widthUnit: "percentage",
        mapProvider: "openStreet",
        divStyles: {},
        inPreviewMode: false
    };

    const renderLeafletMap = (props: LeafletMapProps): ShallowWrapper<LeafletMapProps, any> =>
        shallow(createElement(LeafletMap, props));
    const fullRenderLeafletMap = (props: LeafletMapProps): ReactWrapper<LeafletMapProps, any> =>
        mount(createElement(LeafletMap, props));

    it("renders structure correctly", () => {
        const leafletMap = renderLeafletMap(defaultProps);
        leafletMap.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "pixels"
        });

        expect(leafletMap).toMatchSnapshot();
    });

    it("with pixels renders structure correctly", () => {
        const leafletMap = renderLeafletMap(defaultProps);
        leafletMap.setProps({
            heightUnit: "pixels",
            widthUnit: "pixels"
        });

        expect(leafletMap).toMatchSnapshot();
    });

    it("with percentage of width and height units renders the structure correctly", () => {
        const leafletMap = renderLeafletMap(defaultProps);
        leafletMap.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(leafletMap).toMatchSnapshot();
    });

    it("with percentage of parent units renders the structure correctly", () => {
        const leafletMap = renderLeafletMap(defaultProps);
        leafletMap.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });
        expect(leafletMap).toMatchSnapshot();
    });

    it("without default center Latitude and Longitude sets default center location based on the default configured location", () => {
        const leafletMap = fullRenderLeafletMap(defaultProps);
        leafletMap.setProps({
            fetchingData: false,
            allLocations: []
        });

        expect(leafletMap.state("center")).toEqual({ lat: 51.9066346, lng: 4.4861703 });
    });

    it("un mounts from the dom when unmounted", () => {
        const leafletMap = renderLeafletMap(defaultProps);
        const leafletInstance = leafletMap.instance() as any;
        const componentWillUnmount = spyOn(leafletInstance, "componentWillUnmount").and.callThrough();

        leafletInstance.map = document.createElement("div");
        leafletMap.unmount();

        expect(componentWillUnmount).toHaveBeenCalled();
    });

    it("creates markers from given locations with a url", () => {
        const customProps = {
            ...defaultProps,
            allLocations: [
                { latitude: 40.759011, longitude: -73.9844722, mxObject: undefined, url: "http://dummy.url" }
            ],
            fetchingData: false,
            autoZoom: false
        };
        const leafletMap = fullRenderLeafletMap(defaultProps);
        const leafletMapInstance = leafletMap.instance() as any;
        const createMarkerSpy = spyOn(leafletMapInstance, "renderMarkers").and.callThrough();
        leafletMapInstance.componentWillReceiveProps(customProps);

        expect(createMarkerSpy).toHaveBeenCalledWith(customProps.allLocations);
    });

    it("creates markers from given locations with default icon", () => {
        const customProps = {
            ...defaultProps,
            allLocations: [{ latitude: 40.759011, longitude: -73.9844722, mxObject: undefined }],
            fetchingData: false,
            autoZoom: false
        };
        const leafletMap = fullRenderLeafletMap(defaultProps);
        const leafletMapInstance = leafletMap.instance() as any;
        const createMarkerSpy = spyOn(leafletMapInstance, "renderMarkers").and.callThrough();
        leafletMapInstance.componentWillReceiveProps(customProps);

        expect(createMarkerSpy).toHaveBeenCalledWith(customProps.allLocations);
    });
});
