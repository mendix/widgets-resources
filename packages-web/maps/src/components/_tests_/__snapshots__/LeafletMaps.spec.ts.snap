// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaflet maps renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "37.5px",
      "width": "50px",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-leaflet-maps-alert leaflet-control"
  />
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <div
      className="widget-leaflet-maps"
    />
    <ResizeDetector
      handleHeight={true}
      handleWidth={true}
      nodeType="div"
      onResize={[Function]}
      querySelector={null}
      refreshMode="throttle"
      refreshRate={100}
      skipOnMount={false}
      targetDomEl={null}
    />
  </div>
</div>
`;

exports[`Leaflet maps with percentage of parent units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75%",
      "width": "50%",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-leaflet-maps-alert leaflet-control"
  />
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <div
      className="widget-leaflet-maps"
    />
    <ResizeDetector
      handleHeight={true}
      handleWidth={true}
      nodeType="div"
      onResize={[Function]}
      querySelector={null}
      refreshMode="throttle"
      refreshRate={100}
      skipOnMount={false}
      targetDomEl={null}
    />
  </div>
</div>
`;

exports[`Leaflet maps with percentage of width and height units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "auto",
      "paddingBottom": "37.5%",
      "width": "50%",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-leaflet-maps-alert leaflet-control"
  />
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <div
      className="widget-leaflet-maps"
    />
    <ResizeDetector
      handleHeight={true}
      handleWidth={true}
      nodeType="div"
      onResize={[Function]}
      querySelector={null}
      refreshMode="throttle"
      refreshRate={100}
      skipOnMount={false}
      targetDomEl={null}
    />
  </div>
</div>
`;

exports[`Leaflet maps with pixels renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50px",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-leaflet-maps-alert leaflet-control"
  />
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <div
      className="widget-leaflet-maps"
    />
    <ResizeDetector
      handleHeight={true}
      handleWidth={true}
      nodeType="div"
      onResize={[Function]}
      querySelector={null}
      refreshMode="throttle"
      refreshRate={100}
      skipOnMount={false}
      targetDomEl={null}
    />
  </div>
</div>
`;
